File: python/bestline.py
#!/usr/bin/pypy3
from line import Point
from line import Line

class BestLine:
	def __init__(self, points: list):
		if (len(points) < 2):
			self.numPoints = len(points)
			return
		self.numPoints = 0
		for i in range(0, len(points) - 1):
			for j in range(i + 1, len(points)):
				a = points[i]
				b = points[j]
				l = Line(a, b)
				intersections = 2 + sum(1 for p in points[j+1:] if l.intersects(p))
				if (intersections > self.numPoints):
					self.numPoints = intersections
					self.bestFit = l

File: kotlin/BestLine.kt
class BestLine(points: MutableList<Point>) {
	var numPoints: Int = 0

	init {
		if (points.size < 2) {
			numPoints = points.size
		} else {
			for (i in 0 until points.size - 1) {
				for (j in i + 1 until points.size) {
					val a = points[i]
					val b = points[j]
					val l = Line(a, b)
					val intersections = 2 + points.subList(j + 1, points.size).
							count{l.intersects(it)}
					if (intersections > numPoints) {
						numPoints = intersections
					}
				}
			}
		}
	}

	fun intersections(): Int {
		return numPoints
	}
}
